Sources/HummingbirdRouter/RouteController.swift

```diff
@@ -0,0 +1,106 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the Hummingbird server framework project
+//
+// Copyright (c) 2023-2024 the Hummingbird authors
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See hummingbird/CONTRIBUTORS.txt for the list of Hummingbird authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+import Hummingbird
+
+// MARK: - RouteController
+
+/// A type that defines a body of routes.
+///
+/// You can create custom controllers by declaring types that conform to the `RouteController`
+/// protocol. Implement the required ``RouteController/body`` computed property and provide a
+/// type for the controller's `Context`.
+///
+///     struct UsersController {
+///         typealias Context = BasicRouterRequestContext
+///
+///         var body: some RouterMiddleware<Context> {
+///             RouteGroup("/users") {
+///                 ...
+///             }
+///         }
+///     }
+///
+/// Assemble the controller's body by combining one or more components together to create
+/// a route. By nesting controllers within each other, you can compose large and complex
+/// routes from smaller more mangeable components.
+public protocol RouteController where Context == Body.Context {
```

Comment:
A primary associated type would be nice, so you can filter for `RouteController<MyContextType>`

```suggestion
public protocol RouteController<Context> where Context == Body.Context {
```